components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/pdp/EntitlementEngine.java

```diff
@@ -124,18 +128,63 @@ public static EntitlementEngine getInstance() {
                     }
                 }
             }
-
             return entitlementEngine;
         }
+        if (entitlementEngineLoadingCache == null) {
+            synchronized (lock) {
+                if (entitlementEngineLoadingCache == null) {
+                    entitlementEngineLoadingCache = CacheBuilder.newBuilder().weakValues().expireAfterAccess
+                            (getCacheInterval(), TimeUnit.SECONDS)
+                            .build(new CacheLoader<Integer, EntitlementEngine>() {
+                                @Override
+                                public EntitlementEngine load(Integer key) {
+
+                                    return new EntitlementEngine(key);
+                                }
+                            });
+                }
+            }
+        }
 
-        if (!entitlementEngines.contains(tenantId)) {
+        EntitlementEngine entitleEngine = entitlementEngineLoadingCache.getIfPresent(tenantId);
+        if (entitleEngine == null) {
             synchronized (lock) {
-                if (!entitlementEngines.contains(tenantId)) {
-                    entitlementEngines.put(tenantId, new EntitlementEngine(tenantId));
+                entitleEngine = entitlementEngineLoadingCache.getIfPresent(tenantId);
+                if (entitleEngine == null) {
+                    entitlementEngineLoadingCache.put(tenantId, new EntitlementEngine(tenantId));
+                }
+            }
+        }
+        try {
+            entitleEngine = entitlementEngineLoadingCache.get(tenantId);
+        } catch (ExecutionException e) {
+            log.error("Error while getting the entitle engine for the tenant : " + tenantId);
+        }
+        return entitleEngine;
+    }
+
+    private static long getCacheInterval() {
+
+        Properties properties = EntitlementServiceComponent.getEntitlementConfig().getEngineProperties();
+        String engineCachingInterval = properties.getProperty(PDPConstants
+                .ENTITLEMENT_ENGINE_CACHING_INTERVAL);
+        long entitlementEngineCachingInterval = DEFAULT_ENTITLEMENT_ENGINE_CACHING_INTERVAL;
+        if (engineCachingInterval != null) {
+            try {
+                entitlementEngineCachingInterval = Long.parseLong(engineCachingInterval);
+            } catch (NumberFormatException e) {
+                if (log.isDebugEnabled()) {
+                    log.warn("Invalid value for " + PDPConstants.ENTITLEMENT_ENGINE_CACHING_INTERVAL + ". Using " +
```

Comment:
shouldn't this be log.debug

Reply:
It should be warn log and log.isdebugEnabled should be removed. I have removed it now.